troposphere/static/js/components/modals/instance/InstanceLaunchWizardModal.react.js

```diff
@@ -1,396 +1,457 @@
     },
 
-    getState: function() {
-      return this.state;
+    onAddAttachedScript: function(value) {
+        let attachedScripts = this.state.attachedScripts;
+
+        this.setState({ attachedScripts: [...attachedScripts, value] });
     },
 
-    updateState: function() {
-      if (this.isMounted()) this.setState(this.getState());
+    onRemoveAttachedScript: function(item) {
+        let attachedScripts = this.state.attachedScripts
+            .filter((i) => i != item);
+
+        this.setState({ attachedScripts });
     },
 
-    componentDidMount: function() {
-      stores.ProviderStore.addChangeListener(this.updateState);
-      stores.IdentityStore.addChangeListener(this.updateState);
-      stores.SizeStore.addChangeListener(this.updateState);
-      stores.ProjectStore.addChangeListener(this.updateState);
-      stores.ProjectVolumeStore.addChangeListener(this.updateState);
-      stores.ProjectInstanceStore.addChangeListener(this.updateState);
-      stores.InstanceStore.addChangeListener(this.updateState);
-      stores.ImageVersionStore.addChangeListener(this.updateState);
-      stores.MaintenanceMessageStore.addChangeListener(this.updateState);
-      stores.ScriptStore.addChangeListener(this.updateState);
-      stores.LicenseStore.addChangeListener(this.updateState);
+    onSaveAdvanced: function() {
+        this.viewBasic()
     },
 
-    componentWillUnmount: function() {
-      stores.ProviderStore.removeChangeListener(this.updateState);
-      stores.IdentityStore.removeChangeListener(this.updateState);
-      stores.SizeStore.removeChangeListener(this.updateState);
-      stores.ProjectStore.removeChangeListener(this.updateState);
-      stores.ProjectVolumeStore.removeChangeListener(this.updateState);
-      stores.ProjectInstanceStore.removeChangeListener(this.updateState);
-      stores.InstanceStore.removeChangeListener(this.updateState);
-      stores.ImageVersionStore.removeChangeListener(this.updateState);
-      stores.MaintenanceMessageStore.removeChangeListener(this.updateState);
-      stores.ScriptStore.removeChangeListener(this.updateState);
-      stores.LicenseStore.removeChangeListener(this.updateState);
+    onCancelAdvanced: function() {
+        this.setState({ attachedScripts: [] });
+        this.viewBasic();
     },
 
-    //
-    // Internal Modal Callbacks
-    // ------------------------
-    //
+    canLaunch: function() {
+        var requiredFields = ["project", "identityProvider", "providerSize", "imageVersion", "attachedScripts"];
+        var notFalsy = ((prop) => Boolean(this.state[prop]) != false);
 
-    cancel: function() {
-      this.hide();
+        // instanceName will be null, indicating that it has not been set.
+        // If instanceName equals the empty string, the user has erased the
+        // name, and is trying to launch an instance with no name.
+        return _.every(requiredFields, notFalsy) && this.state.instanceName !== "";
```

Comment:
`canLaunch` is truly the validation location where we should be doing a `.trim()` since we're using it to determine if we have all the necessary _state_ to launch the instance.